src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/ProjectSystem/RazorProjectInfoEndpointPublisher.cs

```diff
@@ -0,0 +1,194 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+using Microsoft.AspNetCore.Razor.LanguageServer.ProjectSystem;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+using Microsoft.CodeAnalysis.Razor.Utilities;
+using Microsoft.VisualStudio.LanguageServer.ContainedLanguage;
+
+namespace Microsoft.VisualStudio.LanguageServerClient.Razor.ProjectSystem;
+
+/// <summary>
+/// Publishes project data (including TagHelper info) discovered OOB to the server via LSP notification
+/// instead of old method of writing a project configuration bin file
+/// </summary>
+[Shared]
+[Export(typeof(RazorProjectInfoEndpointPublisher))]
+internal partial class RazorProjectInfoEndpointPublisher
+{
+    private readonly LSPRequestInvoker _requestInvoker;
+    private readonly IProjectSnapshotManager _projectManager;
+
+    private readonly AsyncBatchingWorkQueue<(IProjectSnapshot Project, bool Removal)> _workQueue;
+
+    private bool _active;
+
+    [ImportingConstructor]
+    public RazorProjectInfoEndpointPublisher(
+        LSPRequestInvoker requestInvoker,
+        IProjectSnapshotManager projectManager
+    )
+    {
+        _requestInvoker = requestInvoker;
+        _projectManager = projectManager;
+
+        _workQueue = new(
+            EnqueueDelay,
+            ProcessBatchAsync,
+            CancellationToken.None);
+    }
+
+    // internal for testing
+    // Delay between publishes to prevent bursts of changes yet still be responsive to changes.
+    internal TimeSpan EnqueueDelay { get; set; } = TimeSpan.FromMilliseconds(250);
+
+    public void StartSending()
+    {
+        _active = true;
+
+        _projectManager.Changed += ProjectManager_Changed;
+
+        var projects = _projectManager.GetProjects();
+        foreach (var project in projects)
+        {
+            if (ProjectWorkspacePublishable(project))
+            {
+                // Don't enqueue project addition as those are unlikely to come through in large batches.
+                // Also ensures that we won't get project removal go through the queue without project addition.
+                ImmediatePublish(project);
+            }
+        }
+    }
+
+    private void ProjectManager_Changed(object sender, ProjectChangeEventArgs args)
+    {
+        // Don't do any work if the solution is closing
+        if (args.SolutionIsClosing)
+        {
+            return;
+        }
+
+        switch (args.Kind)
+        {
+            case ProjectChangeKind.ProjectChanged:
+            case ProjectChangeKind.DocumentChanged:
```

Comment:
Should we handle DocumentChanged? if the comment in the other file is to be believed, it happens on every keystroke, which is probably a bit much.

Reply:
You are right, and yes - it seems to happen on every keystroke. Looks like old code used it only for the first notification if file didn't exist yet. Removed it.